[expo-media-library][ios] Add `presentPermissionsPickerAsync` (#10662)
expo/expo: docs/pages/versions/unversioned/sdk/media-library.md

A promise that resolves to an object of type [CameraRollPermissionResponse](#medialibrarycamerarollpermissionresponse).

### `MediaLibrary.presentPermissionsPickerAsync()`

**Available only on iOS >= 14.** Allows the user to update the assets which your app has access to. The system modal is displayed only if a user originally selected the `limited` access to the media library.

#### Returns

A promise that resolves to void if prompt will be displayed correctly.

> > **Note:** This method doesn't inform you if the user changes permitted assets. To check this you need to subscribe for updates in the user's media library using [MediaLibrary.addListener(listener)](#medialibraryaddlistenerlistener). If `shouldFullyReload` is set to true, the user changed the permissions.

Author: ```suggestion
> **Note:** This method doesn't inform you if the user changes which assets your app has access to. For that information, you need to subscribe for updates to the user's media library using [MediaLibrary.addListener(listener)](#medialibraryaddlistenerlistener). If `hasIncrementalChanges` is `false`, the user changed their permissions.
```

Ohh. I think this should look like this :D 

If the user has changed permissions the `hasIncrementalChanges` will be set to `false`.

> > **Note:** This method doesn't inform you if the user changes permitted assets. To check this you need to subscribe for updates in the user's media library using [MediaLibrary.addListener(listener)](#medialibraryaddlistenerlistener). If `shouldFullyReload` is set to true, the user changed the permissions.

Reviewer: aha! okay that makes more sense 😄

A promise that resolves to an object of type [CameraRollPermissionResponse](#medialibrarycamerarollpermissionresponse).

### `MediaLibrary.presentPermissionsPickerAsync()`

> **Available only on iOS >= 14.** Allows the user to update the assets which your app has access to. It displays a system modal only if permissions were `granted` and the `accessPrivileges` is equal to `limited`.

Reviewer 2: ```suggestion
**Available only on iOS >= 14.** Allows the user to update the assets which your app has access to. The system modal is displayed only if a user originally selected the `limited` access to the media library.
```
I would go with sth like this 🤔 
I guess there's still a better and simpler way to describe the actual behaviour than what I'm suggesting 😅

A promise that resolves to an object of type [CameraRollPermissionResponse](#medialibrarycamerarollpermissionresponse).

### `MediaLibrary.presentLimitedLibraryPickerAsync()`

> **Available only on iOS >= 14.** Allows the user to update the assets which your app will have access to. This method will display a system prompt only if permissions were `granted` and the `accessPrivileges` is equal to `limited`.

Reviewer 3: ```suggestion
**Available only on iOS >= 14.** Allows the user to update the assets which your app has access to. It displays a system modal only if permissions were `granted` and the `accessPrivileges` is equal to `limited`.
```

#### Arguments

- **listener (_function_)** -- A callback that is fired when any assets have been inserted or deleted from the library, or when the user changes which assets they're allowing access to. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object containing following keys:

>   - **hasIncrementalChanges (_boolean_)** -- Whether the media library's changes could be described by the `incremental changes`. `true` indicates the changes are described by the `insertedAssets`, `deletedAssets` and `updatedAssets` values. `false` indicates that the scope of changes is too large and a full assets reload should be performed by the application, `incremental changes` could not be provided in this case (eg. a user has changed the access to the individual assets in the media library).

Author: So when the user change permissions the `hasIncrementalChanges` key will be set to false. This is an example that shows when the user needs to reload all assets.

>   - **hasIncrementalChanges (_boolean_)** -- Whether the media library's changes could be described by the `incremental changes`. `true` indicates the changes are described by the `insertedAssets`, `deletedAssets` and `updatedAssets` values. `false` indicates that the scope of changes is too large and a full assets reload should be performed by the application, `incremental changes` could not be provided in this case (eg. a user has changed the access to the individual assets in the media library).

Reviewer: Got it, I understand now. With [this change](https://github.com/expo/expo/pull/10662#discussion_r522863842), that will be clear now in the docs 👍

#### Arguments

- **listener (_function_)** -- A callback that is fired when any assets have been inserted or deleted from the library, or when the user changes which assets they're allowing access to. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object containing following keys:

> 
>   - **hasIncrementalChanges (_boolean_)** -- Whether the changes could be described by the incremental changes. `true` if the changes are described by the removed/inserted/changed details. `false` indicates that the scope of changes were too large and UI clients should do a full reload, incremental changes could not be provided (for example, if the user has changed access to individual assets in their library).

Reviewer 2: ```suggestion
  - **hasIncrementalChanges (_boolean_)** -- Whether the media library's changes could be described by the `incremental changes`. `true` indicates the changes are described by the `insertedAssets`, `deletedAssets` and `updatedAssets` values. `false` indicates that the scope of changes is too large and a full assets reload should be performed by the application, `incremental changes` could not be provided in this case (eg. a user has changed the access to the individual assets in the media library).
```

>   Presents only if `hasIncrementalChanges` is `true`:

Reviewer 2: ```suggestion
  Available only if `hasIncrementalChanges` is `true`:
```

#### Arguments

- **listener (_function_)** -- A callback that is fired when any assets have been inserted or deleted from the library, or when the user changes which assets they're allowing access to. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object containing following keys:

>   - **assetPermissionsChanged (_boolean_)** -- Flag indicating a permissions change occurred in the user's limited library selection. If `true`, the user has changed access to individual assets in their library, and you should reload all used assets to comply with the new permissions.

Reviewer 2: ```suggestion
  - **assetsPermissionsChanged (_boolean_)** -- Flag indicating a permissions change occurred in the user's limited library selection. If `true`, the user has changed access to individual assets in their library, and you should reload all used assets to comply with the new permissions.
```
Yeah, I like this name more then the previous one 😉 
I suggest going with the plural form as there's number of assets rather than single one 🤔

#### Arguments

> - **listener (_function_)** -- A callback that is called when any assets have been inserted or deleted from the library. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object that contains following keys:

Reviewer: We should updated this description now that the listener is also called when the permissions change. Maybe something like
```suggestion
- **listener (_function_)** -- A callback that is fired when any assets have been inserted or deleted from the library, or when the user changes which assets they're allowing access to. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object containing following keys:
```

>   - **shouldFullyReload (_boolean_)** -- Whether to all assets reloads is needed. When the user changes the permissions, this value will be set to `true` and you should reload all used assets.

Reviewer 2: This name is very confusing without reading the whole docs and it suggest specific scenario that should be performed 😅 I'm not a fan of this 😉 

Let's see what is actually happening. The idea is to inform that the selection has changed yes? So what about naming it that way?
Names coming to my mind are: `assetsPermissionsChanged`, `permittedAssetsSelectionChanged` 🤔 

```suggestion
  - **assetsPermissionsChanged (_boolean_)** -- Flag indicating the permissions change in the `user's limited library selection`. If it's `true` this means that the user changed the access to the individual assets in the library. If this happens you should reload all used assets to resemble the actual access permissions.
```

cc @cruzach

Reviewer: I agree, `assetsPermissionsChanged` makes it clear what's happening behind the scenes, rather than just us telling the user what to do (fully reload)
```suggestion
  - ** assetPermissionsChanged (_boolean_)** -- Flag indicating a permissions change occurred in the user's limited library selection. If `true`, the user has changed access to individual assets in their library, and you should reload all used assets to comply with the new permissions.
```
